File: 프로그래머스/2/131704. 택배상자/택배상자.cpp
#include <string>
#include <vector>

using namespace std;

int solution(vector<int> order) {
    int answer = 0;
    	vector<int> temp;

	int want = 0;
	int num = 1;

	while (true)
	{
		if (num == order[want])
		{
			want++;
			num++;
			answer++;
		}
		else
		{
			if (temp.empty())
			{
				temp.push_back(num);
				num++;
			}
			else
			{
				if (temp.back() == order[want])
				{
					temp.pop_back();
					answer++;
					want++;
				}
				else
				{
					temp.push_back(num);
					num++;
				}
			}
		}
		if (num > order.size()) break;
	}

	while (!temp.empty())
	{
		if (temp.back() == order[want])
		{
			temp.pop_back();
			answer++;
			want++;
		}
		else break;
	}
    return answer;
}

File: 프로그래머스/2/131704. 택배상자/택배상자.cs
using System;
using System.Linq;
using System.Collections.Generic;
public class Solution {
    public int solution(int[] order) {
        int answer = 0;
        Stack<int> stk = new Stack<int>();
int k = 1;

for (int i = 0; i < order.Length; i++)
{
    bool chk = true;

    while (order[i] != k)
    {
        if(stk.Count() != 0)
        {
            if(stk.Peek() == order[i])
            {
                stk.Pop();
                answer++;
                break;
            }
        }

        if (k > order.Length)
        {
            chk = false;
            break;
        }
        stk.Push(k++);
    }
    if (!chk) break;

    if (order[i] == k)
    {
        answer++;
        k++;
    }

}
        return answer;
    }
}